lua/dap/session.lua

```diff
@@ -560,17 +560,31 @@ function Session:handle_body(body)
         log.warn('No event handler for ', decoded)
       end
     end)
-  elseif decoded.type == 'request' and decoded.command == 'runInTerminal' then
-    self:run_in_terminal(decoded)
+  elseif decoded.type == 'request' then
+    local handler = self.handlers.reverse_requests[decoded.command]
+    if handler then
+      handler(self, decoded)
+    end
+    log.warn('No handler for reverse request', decoded)
   else
     log.warn('Received unexpected message', decoded)
   end
 end
 
 
-local function session_defaults(opts)
+local default_reverse_request_handlers = {
+  runInTerminal = run_in_terminal
+}
+
+
+local function session_defaults(adapter, opts)
```

Comment:
Having the adapter here will be useful as I think another nice-to-have for vscode-js-debug would be to have a custom onExit for the session. This change should make that trivial to implement.

Reply:
Could you elaborate why that would be useful - and why the existing `dap.listeners` for the exit/terminated events wouldn't be sufficient?

Reply:
The problem is that for javascript, you need to run a command which starts the server and outputs a port and this is the port you need to use in the adapter above. But at that point there is no session yet, so you cannot attach the onexit (to stop the original process) on the session

https://github.com/entropitor/dotfiles/blob/06aa5ea67fad9be8e828ec0cafbada7b186df82b/tag-vim/config/nvim/lua/entropitor/dap.lua#L9
https://github.com/entropitor/dotfiles/blob/06aa5ea67fad9be8e828ec0cafbada7b186df82b/tag-vim/config/nvim/lua/entropitor/dap/dap-js.lua#L88

Reply:
You can register listeners at any time and registering them doesn't require a session.
The `connect` method could do something like this:


```lua
dap.listeners.before['event_terminated']['your-unique-plugin-id'] = function(session, body)
  -- stop process here

  dap.listeners.before['event_terminated']['your-unique-plugin-id'] = nil
end
```